File: java/CloneGraph.java
import java.util.HashMap;
import java.util.HashSet;

public class CloneGraph {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        HashMap<UndirectedGraphNode, UndirectedGraphNode> map 
            = new HashMap<UndirectedGraphNode, UndirectedGraphNode>();
        map.put((UndirectedGraphNode) null, (UndirectedGraphNode) null);
        buildMap(node, map);
        HashSet<UndirectedGraphNode> set = new HashSet<UndirectedGraphNode>();
        set.add((UndirectedGraphNode) null);
        buildGraph(node, map, set);
        return map.get(node);
    }

    private void buildMap(UndirectedGraphNode node, 
            HashMap<UndirectedGraphNode, UndirectedGraphNode> map) {
        if (map.containsKey(node)) {
            return;
        }
        map.put(node, new UndirectedGraphNode(node.label));
        for (UndirectedGraphNode next : node.neighbors) {
            buildMap(next, map);
        }
    }

    private void buildGraph(UndirectedGraphNode node, 
            HashMap<UndirectedGraphNode, UndirectedGraphNode> map, 
            HashSet<UndirectedGraphNode> set) {
        if (set.contains(node)) {
            return;
        }
        set.add(node);
        for (UndirectedGraphNode next : node.neighbors) {
            map.get(node).neighbors.add(map.get(next));
        }
        for (UndirectedGraphNode next : node.neighbors) {
            buildGraph(next, map, set);
        }
    }
}

File: python/clone_graph.py
#/usr/bin/env python3

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None
        track = dict()
        track[node] = UndirectedGraphNode(node.label)
        queue = list()
        queue.append(node)
        while queue:
            curr = queue.pop(0)
            for neighbor in curr.neighbors:
                if neighbor not in track:
                    track[neighbor] = UndirectedGraphNode(neighbor.label)
                    queue.append(neighbor)
            for neighbor in curr.neighbors:
                track[curr].neighbors.append(track[neighbor])
        return track[node]

def main():
    solver = Solution()

if __name__ == '__main__':
    main()
